File: my_py_pkg/my_py_pkg/number_counter.py
import rclpy
from rclpy.node import Node
from example_interfaces.msg import Int16


class NumberCounter(Node):

    def __init__(self):
        super().__init__("number_counter")
        self.get_logger().info("Hello from " + self.get_name())
        self.counter_ = 0
        self.publisher = self.create_publisher(Int16, "number_counter", 10)
        self.subscription = self.create_subscription(Int16, "numbers", self.callback, 10)

    def callback(self, msg):
        self.counter_ += 1
        data = Int16()
        data.data = self.counter_
        self.publisher.publish(data)


def main(args=None):
    rclpy.init(args=args)
    node = NumberCounter()
    
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: my_cpp_pkg/src/number_counter.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/int16.hpp"

class NumberCounter : public rclcpp::Node
{
public:
    NumberCounter() : Node("number_counter")
    {
        RCLCPP_INFO(this->get_logger(), "Hello from %s", this->get_name());
        counter = 0;

        subscription = this->create_subscription<example_interfaces::msg::Int16>("numbers", 10,
            std::bind(&NumberCounter::myCallback, this, std::placeholders::_1));
            // std::placeholders::_1 needs because the callback has one parameter. If it would have more, use _2 or _3, etc instead of _1

        publisher = this->create_publisher<example_interfaces::msg::Int16>("number_counter", 10);

    }

private:
    rclcpp::Subscription<example_interfaces::msg::Int16>::SharedPtr subscription;
    rclcpp::Publisher<example_interfaces::msg::Int16>::SharedPtr publisher;

    void myCallback(const example_interfaces::msg::Int16::SharedPtr msg)
    {
        counter++;
        auto data = example_interfaces::msg::Int16();
        data.data = counter;
        publisher->publish(data);
        RCLCPP_INFO(this->get_logger(), "Got number %d", msg->data);
    }

    int counter;

};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NumberCounter>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
